### PR TITLE
Add NetCDF chunking support for optimized I/O performance

### PR DESCRIPTION
**Pull request by siligam**
## Overview

This PR implements internal NetCDF chunking functionality to optimize I/O performance when reading and writing climate data files. The implementation is inspired by the [dynamic_chunks](https://github.com/jbusecke/dynamic_chunks) library and provides flexible chunking strategies that can be configured globally or per-variable.

## Motivation

NetCDF-4 files can be internally "chunked" to improve I/O performance:
- **Better read performance**: Reading subsets of data becomes faster when chunks align with access patterns
- **Compression efficiency**: Chunked data can be compressed more effectively  
- **Parallel I/O**: Chunked files enable better parallel read/write operations
- **Optimal for time-series**: Climate data is often accessed along the time dimension, so chunking along time improves performance

### New Files

1. **`src/pycmor/std_lib/chunking.py`** (470 lines)
   - Three chunking algorithms:
     - `calculate_chunks_simple()`: Fast, time-dimension-focused (default)
     - `calculate_chunks_even_divisor()`: Finds chunks that evenly divide dimensions
     - `calculate_chunks_iterative()`: Iteratively scales chunks to match target size
   - `get_encoding_with_chunks()`: Generates xarray encoding with chunks and compression
   - `get_memory_size()`: Estimates memory size for chunk configurations

2. **`tests/unit/test_chunking.py`** (290 lines)
   - 13 comprehensive tests covering all algorithms
   - Tests for edge cases, compression settings, and custom configurations
   - All tests passing ✓

3. **`doc/netcdf_chunking.rst`**
   - Complete user guide with examples
   - Algorithm comparisons and use cases
   - Performance tips and troubleshooting

4. **`examples/chunking_example.yaml`**
   - Example configuration showing global and per-rule settings
   - Multiple use cases demonstrated

### Modified Files

1. **`src/pycmor/core/config.py`**
   - Added 7 new configuration options:
     - `netcdf_enable_chunking` (default: yes)
     - `netcdf_chunk_algorithm` (default: simple, choices: simple/even_divisor/iterative)
     - `netcdf_chunk_size` (default: 100MB)
     - `netcdf_chunk_tolerance` (default: 0.5)
     - `netcdf_chunk_prefer_time` (default: yes)
     - `netcdf_compression_level` (default: 4, range: 1-9)
     - `netcdf_enable_compression` (default: yes)

2. **`src/pycmor/std_lib/files.py`**
   - Added `_calculate_netcdf_chunks()` helper function
   - Integrated chunking into all `to_netcdf()` and `save_mfdataset()` calls
   - Supports per-rule override of chunking settings
   - Graceful error handling (falls back to no chunking on failure)

3. **`doc/index.rst`**
   - Added chunking guide to documentation table of contents

## Features

✅ **Flexible Configuration**: Global and per-rule settings via YAML  
✅ **Multiple Algorithms**: Choose based on your use case  
✅ **Automatic Compression**: Built-in zlib compression (level 1-9)  
✅ **Time-Dimension Optimization**: Optimized for climate data access patterns  
✅ **Error Handling**: Graceful fallback if chunking fails  
✅ **Comprehensive Testing**: 13 tests, all passing  
✅ **Detailed Documentation**: User guide with examples  
✅ **Backward Compatible**: Can be disabled if needed  

### Global Configuration

In `~/.config/pycmor/pycmor.yaml`:

```yaml
# Enable chunking with default settings
netcdf_enable_chunking: yes
netcdf_chunk_algorithm: simple
netcdf_chunk_size: 100MB
netcdf_chunk_prefer_time: yes
netcdf_compression_level: 4
```

### Per-Rule Override

In your processing configuration:

```yaml
rules:
  - model_variable: temperature
    cmor_variable: tas
    cmor_table: CMIP6_Amon.json
    # Override chunking for this variable
    netcdf_chunk_algorithm: even_divisor
    netcdf_chunk_size: 50MB
    netcdf_compression_level: 6
```

### Programmatic Usage

```python
from pycmor.std_lib.chunking import calculate_chunks_simple, get_encoding_with_chunks
import xarray as xr

ds = xr.open_dataset('input.nc')
chunks = calculate_chunks_simple(ds, target_chunk_size='100MB')
encoding = get_encoding_with_chunks(ds, chunks=chunks, compression_level=4)
ds.to_netcdf('output.nc', encoding=encoding)
```

## Default Behavior

With default settings:
- Chunking is **enabled**
- Uses **simple** algorithm
- Target chunk size: **100MB**
- Prefers **time-dimension** chunking
- Compression level: **4** (balanced)
- Compression **enabled**

This provides good performance for most climate data use cases without requiring configuration.

## Testing

All tests pass:

```bash
pytest tests/unit/test_chunking.py -v
# 13 passed, 7 warnings in 0.78s
```

## Performance Benefits

- **Faster I/O**: Optimized chunk sizes improve read/write performance
- **Better Compression**: Chunked data compresses more efficiently (50-70% reduction typical)
- **Reduced File Size**: Compression can significantly reduce storage requirements
- **Parallel Access**: Chunked files enable better parallel I/O operations

## Breaking Changes

None. Chunking is enabled by default but can be disabled:

```yaml
netcdf_enable_chunking: no
```

## Documentation

Complete documentation added at `doc/netcdf_chunking.rst` including:
- Configuration options
- Algorithm descriptions
- Performance tips
- Troubleshooting guide
- Examples for different use cases

## Future Enhancements

Potential improvements for future versions:
- Adaptive chunking based on data access patterns
- Chunk caching for similar datasets
- Performance profiling tools
- Additional chunking strategies

## References

- [NetCDF-4 Chunking Guide](https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_perf_chunking.html)
- [Dynamic Chunks Library](https://github.com/jbusecke/dynamic_chunks) (inspiration)
- [Xarray I/O Documentation](https://docs.xarray.dev/en/stable/user-guide/io.html)

## Checklist

- [x] Code follows project style guidelines
- [x] Tests added and passing
- [x] Documentation updated
- [x] Examples provided
- [x] Backward compatible
- [x] No breaking changes

### CODE REVIEW

File: doc/netcdf_chunking.rst [hunk 1]
```diff
@@ -0,0 +1,362 @@
+============================
+NetCDF Chunking Guide
```
Comment (mandresm) [inline doc/netcdf_chunking.rst:2]: Screenshot 2025-11-12 at 16 26 59

```suggestion
Usage: NetCDF Chunking Guide
```

File: doc/netcdf_chunking.rst [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+============================
+
+Overview
+========
+
+pycmor now supports internal NetCDF chunking to optimize I/O performance when reading and writing climate data files. Chunking divides the data into smaller blocks that can be read and written more efficiently, especially for datasets with specific access patterns.
```
Comment (mandresm) [inline doc/netcdf_chunking.rst]:
```suggestion
`pycmor` supports internal NetCDF chunking to optimize I/O performance when reading and writing climate data files. Chunking divides the data into smaller blocks that can be read and written more efficiently, especially for datasets with specific access patterns.
```

File: doc/netcdf_chunking.rst [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+   # Get encoding with chunks and compression
+   encoding = get_encoding_with_chunks(
+       ds,
+       chunks=chunks,
+       compression_level=4,
+       enable_compression=True
+   )
+
+   # Save with custom encoding
+   ds.to_netcdf('output.nc', encoding=encoding)
```
Comment (mandresm) [inline doc/netcdf_chunking.rst:392]: Great docu!

File: src/pycmor/std_lib/files.py [hunk 4]
```diff
@@ -374,6 +380,85 @@ def _save_dataset_with_native_timespan(
...
+                ds,
+                target_chunk_size=chunk_size,
+                size_tolerance=chunk_tolerance,
+            )
+        elif chunk_algorithm == "iterative":
+            chunks = calculate_chunks_iterative(
+                ds,
+                target_chunk_size=chunk_size,
+                size_tolerance=chunk_tolerance,
+            )
```
Comment (mandresm) [inline src/pycmor/std_lib/files.py]: One option to make this code more compact would be to define a `chunk_functions` dictionary:

```python
chunk_functions = {
    "simple": calculate_chunks_simple,
    "even_divisor": calculate_chunks_even_divisor,
    "iterative": calculate_chunks_iterative,
}
```

Then you can call the function with:

```python
            chunks = chunk_functions[chunk_algorithm](
                ds,
                target_chunk_size=chunk_size,
                prefer_time_chunking=prefer_time,
            )
```

For the warning you can always do:

```python
if chunk_algorithm not in chunk_functions:
    logger.warning(f"Unknown chunk algorithm: {chunk_algorithm}, using simple")
    chunk_algorithm = "simple"
```

**Review by mandresm — CHANGES_REQUESTED** [review]
Hi @siligam, is looking great already, but I made a couple of suggestions.

I still have a couple of files to review but I will continue tomorrow.

**Review by siligam — COMMENTED** [review]
updated all suggested changes

File: examples/chunking_example.yaml
Comment (pgierz) [inline examples/chunking_example.yaml:1]: I think there is some section missing here? Maybe it needs `inherit`, or globally in `pycmor`? I am fairly sure we did not have top level keys.

File: src/pycmor/std_lib/files.py [hunk 5]
```diff
@@ -374,6 +380,72 @@ def _save_dataset_with_native_timespan(
...
+    if not enable_chunking:
+        return {}
+
+    # Get chunking configuration
+    chunk_algorithm = rule._pycmor_cfg("netcdf_chunk_algorithm")
+    chunk_size = rule._pycmor_cfg("netcdf_chunk_size")
+    chunk_tolerance = rule._pycmor_cfg("netcdf_chunk_tolerance")
+    prefer_time = rule._pycmor_cfg("netcdf_chunk_prefer_time")
+    compression_level = rule._pycmor_cfg("netcdf_compression_level")
+    enable_compression = rule._pycmor_cfg("netcdf_enable_compression")
```
Comment (pgierz) [inline src/pycmor/std_lib/files.py:412]: It would be cleaner to get this from an inherit block, then it is on the rule directly.

File: doc/netcdf_chunking.rst [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+- **Parallel I/O**: Chunked files enable better parallel read/write operations
+- **Optimal for time-series**: Climate data is often accessed along the time dimension, so chunking along time improves performance
+
+Configuration Options
+======================
+
+Global Configuration
+--------------------
+
+You can configure chunking globally in your pycmor configuration file (e.g., ``~/.config/pycmor/pycmor.yaml``):
```
Comment (pgierz) [inline doc/netcdf_chunking.rst]: I think this belongs in an inherit block

**Review by pgierz — COMMENTED** [review]
Looks good, but I have a thought: we put the control over chunking behavior for this into the pycmor `config` (that is nice, it gets an automatic control via environment variables), but maybe this belongs to a global inherit that is passed down to the rules?

I am not sure what is better, to be honest.

File: examples/chunking_example.yaml
Comment (siligam) [reply examples/chunking_example.yaml:1]: I will change that to `inherit` block

**Review by pgierz — APPROVED** [review]
